consensus/datasource/consensus_data_source.go

```diff
@@ -0,0 +1,78 @@
+package datasource
+
+import (
+	"context"
+	"errors"
+	syncmap "sync"
+	"sync/atomic"
+
+	"github.com/NethermindEth/juno/consensus/driver"
+	"github.com/NethermindEth/juno/consensus/p2p"
+	"github.com/NethermindEth/juno/consensus/proposer"
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/sync"
+)
+
+const maxCommitHistory = 1024 // TODO: make this configurable
+
+type consensusDataSource[V types.Hashable[H], H types.Hash, A types.Addr] struct {
+	commitListener driver.CommitListener[V, H, A]
+	proposer       proposer.Proposer[V, H]
+	cache          syncmap.Map
+	latest         atomic.Uint64
+}
+
+func New[V types.Hashable[H], H types.Hash, A types.Addr](
+	commitListener driver.CommitListener[V, H, A],
+	proposer proposer.Proposer[V, H],
+	p2p p2p.P2P[V, H, A],
+) *consensusDataSource[V, H, A] {
+	return &consensusDataSource[V, H, A]{
+		commitListener: commitListener,
+		proposer:       proposer,
+		cache:          syncmap.Map{},
```

Comment:
This `Map` type is optimized for "ever growing" cache and when writing/overwriting disjoint sets of key at the same time and multiple reads.

Given the use of `consensusDataSource`, it is meant to write a key, and delete another key. The safety mechanisms come out of the need for reading later on, we want to avoid race condition.

It looks that this data type we are looking for here is a queue (of size maxCommitHistory) with constant look up time. What would be the best implementation for this 🤔

Reply:
Yes, I also plan to work on this after this PR, but it needs to change more on how Synchronizer works.